netperf_tool/src/client_node.hpp

```diff
@@ -0,0 +1,95 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CLIENT_NODE_HPP_
+#define CLIENT_NODE_HPP_
+
+#include <chrono>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "netperf_tool_interfaces/msg/bytes.hpp"
+#include "netperf_tool_interfaces/srv/get_results.hpp"
+
+#include "rclcpp/executors.hpp"
+#include "rclcpp/node.hpp"
+#include "rclcpp/serialization.hpp"
+#include "rclcpp/serialized_message.hpp"
+
+namespace netperf_tool
+{
+struct ClientCollectedInfo
+{
+  std::vector<size_t> message_ids;
+  std::vector<std::chrono::time_point<std::chrono::system_clock>> message_published_times;
+  std::vector<size_t> message_sizes;
+};
```

Comment:
Instead of a struct with three vectors, why not a vector of structs with three entries? Do you ever expect these vectors to have different lengths?

Reply:
> Do you ever expect these vectors to have different lengths?

No, they are always the same length.

> Instead of a struct with three vectors, why not a vector of structs with three entries?

Both are fine IMO.
In `ServerCollectedInfo`, it was nice to have an struct of arrays to calculate the statistics easily, though I can also do that with the array of structs
Here, I imitated what `ServerCollectedInfo` uses.

Actually, I think I'm not using `ClientCollectedInfo` for anything.
The idea of collecting everything is that I could generate a csv with the raw data from it, which I think it's nice, but I haven't added that.

Same for `ServerCollectedInfo`, if we don't want to have the raw data available, I can just calculate the running statistics instead of collection all the data.

But I think that having the possibility to export the raw data to a csv in the future is nice.

Reply:
After applying https://github.com/ros2/ros2cli/pull/720/commits/4cb873ac81d49b34edb5034efc32911fd20bc9ab, I think that this could maybe be:

```cpp
{
  size_t n_messages_published;
  size_t serialized_message_size;
  std::vector<std::chrono::time_point<std::chrono::system_clock>> message_published_times;
};
```

as the messages id are `[0, n_messages_published - 1]` and `serialized_message_size` is a constant.

In the server side, I think it's nice to collect everything:
- It's worth to collect the ids as messages can be lost, to have the complete raw data available.
- It's worth to collect the message size, to be ready for clients doing strange things (though it shouldn't actually change).

I don't mind between array of structs or struct of arrays in that case.
Let me know what you think.

Reply:
Just thinking from the perspective of data processing, i.e. "DataFrames" which a pervasive idea in things like `pandas`, the array of structs is more natural to me. With each struct being a single record in time. Especially since the data within the struct is correlated, i.e. `message_ids[x]` is associated with `message_published_times[x]`.

Reply:
Done, see https://github.com/ros2/ros2cli/pull/720/commits/53a16a0d960a83c98284dee94d6e51c0a8756e5a and https://github.com/ros2/ros2cli/pull/720/commits/e9212d6fd5cf30b4cf9d7daa111ab29f8be8d27a.